### PR TITLE
Added ability to transfer duplicates

### PR DESCRIPTION
**Pull request by beeedy**
Can enable (disabled  by default) a feature where the bot automatically transfers duplicate pokemon. A maximum lvl is also defined where the bot wil not transfer pokemon over a specific lvl, even if a duplicate. A pokemon's "lvl" can be specified as it's cp, iv, cp+iv, or cp*iv.

### CODE REVIEW

File: pgoapi/pgoapi.py [hunk 1]
```diff
@@ -393,6 +397,36 @@ def cleanup_pokemon(self, inventory_items=None):
                             self.log.info("Failed to release Pokemon %s", pokemon)
                         sleep(3)
 
+        if self.RELEASE_DUPLICATES:
+            for pokemons in caught_pokemon.values():
+                if len(pokemons) > MIN_SIMILAR_POKEMON:
```
Comment (joychugh) [inline pgoapi/pgoapi.py]: we can try to run this code in the same loop at the one above it. can you make it a part of that?

**Comment by beeedy** [comment]
Will attempt to combine loops

File: pgoapi/pgoapi.py [hunk 2]
```diff
@@ -380,8 +384,33 @@ def cleanup_pokemon(self, inventory_items=None):
             if len(pokemons) > MIN_SIMILAR_POKEMON:
                 pokemons = sorted(pokemons, key=lambda x: (x.iv, x.cp), reverse=True)
                 # keep the first pokemon....
+                last_pokemon = pokemons[0]
                 for pokemon in pokemons[MIN_SIMILAR_POKEMON:]:
```
Comment (aandromeda) [inline pgoapi/pgoapi.py]: This should be `for pokemon in pokemons[self.MIN_SIMILAR_POKEMON:]:`

File: pgoapi/pgoapi.py [hunk 3]
```diff
@@ -380,8 +384,33 @@ def cleanup_pokemon(self, inventory_items=None):
...
+                        self.release_pokemon(pokemon_id = pokemon.id)
+                        release_res = self.call()['responses']['RELEASE_POKEMON']
+                        status = release_res.get('result', -1)
+                        if status == 1:
+                            self.log.info("Successfully Released Pokemon %s", pokemon)
+                        else:
+                            self.log.debug("Failed to release pokemon %s, %s", pokemon, release_res)
+                            self.log.info("Failed to release Pokemon %s", pokemon)
+                        sleep(3)
+                    elif pokemon.iv < self.MIN_KEEP_IV and pokemon.cp < self.KEEP_CP_OVER and pokemon.is_valid_pokemon():
```
Comment (aandromeda) [inline pgoapi/pgoapi.py]: I don't think the logic is working correctly, it always goes through this elif, never through the other ones, and ends up throwing the higher CP ones sometimes.

Comment (aandromeda) [inline pgoapi/pgoapi.py]: I don't think the logic is working correctly, it always goes through this elif (line 413), never through the other ones, and ends up throwing the higher CP ones sometimes.

File: pgoapi/pgoapi.py (hunk 2 above)
Comment (beeedy) [inline pgoapi/pgoapi.py]: I believe at the time of this PR MIN_SIMILAR_POKEMON was still a hard coded value and not changeable through the config. The line in question wasn't even part of my change set

File: pgoapi/pgoapi.py (hunk 3 above)
Comment (beeedy) [inline pgoapi/pgoapi.py]: I have been testing this for a while now and the logic all seems to be correct. Please make sure you have the config file setup correctly. I did just notice that the print statement makes it appear that the bot is transferring the wrong pokemon at times, but be assured it always is releasing the lesser of the two. I will fix this print statement ASAP

**Comment by beeedy** [comment]
The printout that caused a scare thinking the bot would throw away high lvl pokemon has been fixed! It never was actually transferring them, it was simply a printout error

**Comment by albertonoys** [comment]
I hope this can be merged soon, it's a great feature to have. Thank you for the contribution.

**Comment by beeedy** [comment]
Thank you @albertonoys!

I am going to attempt to fix all conflicts with this and the development branch. There has been a lot of work recently with how to transfer pokemon and I feel that may be why this has not been pulled in yet. Most of the features in this can by synthesized through other means but I personally do not think they are as user friendly. One benefit of this system is to be able to customize how two pokemon are compared to each other, in regards to CP and IV. At the very least I would like that change to be brought in.

**Comment by j-e-k** [comment]
Are you still working on fixing conflicts? Let me know as I would like to get this into develop

**Comment by beeedy** [comment]
I am working on fixing some conflicts but will attempt to get them finished up tonight still. go ahead and close this PR and I will reopen it if you are having problems since the name switch

**Comment by j-e-k** [comment]
Ya name switch has caused a bunch of oddities. Reopen the PR when you are ready for me to take a look and I will get it merged in.
Thanks! 😄

**Comment by beeedy** [comment]
New PR: https://github.com/j-e-k/poketrainer/pull/141

**Pull request closed without merging**